### PR TITLE
speed up intersections

### PR DESCRIPTION
**Pull request by rozchlastywacz**
how it was:
```External
sagittal time: 17.839640617370605 s
coronal time: 12.245706558227539 s
External elapsed: 30.128231048583984
CTV_6000
sagittal time: 0.4264366626739502 s
coronal time: 0.5313541889190674 s
CTV_6000 elapsed: 0.9587903022766113
BrainStem
sagittal time: 0.14860320091247559 s
coronal time: 0.1306602954864502 s
BrainStem elapsed: 0.2802610397338867
Chiasm
sagittal time: 0.027945995330810547 s
coronal time: 0.016968965530395508 s
Chiasm elapsed: 0.044914960861206055
LacrimalGland_L
sagittal time: 0.015944957733154297 s
coronal time: 0.02394270896911621 s
LacrimalGland_L elapsed: 0.03988766670227051
LacrimalGland_R
sagittal time: 0.01400899887084961 s
coronal time: 0.020938634872436523 s
LacrimalGland_R elapsed: 0.03494763374328613
Lens_L
sagittal time: 0.010988712310791016 s
coronal time: 0.010943889617919922 s
Lens_L elapsed: 0.021932601928710938
Lens_R
sagittal time: 0.011013507843017578 s
coronal time: 0.008932352066040039 s
Lens_R elapsed: 0.019945859909057617
OpticNerve_L
sagittal time: 0.029902935028076172 s
coronal time: 0.05588555335998535 s
OpticNerve_L elapsed: 0.08578848838806152
OpticNerve_R
sagittal time: 0.04787397384643555 s
coronal time: 0.054892539978027344 s
OpticNerve_R elapsed: 0.10276651382446289
CTV_5000
sagittal time: 0.9807066917419434 s
coronal time: 1.2633070945739746 s
CTV_5000 elapsed: 2.2480032444000244
Brain
sagittal time: 2.2313084602355957 s
coronal time: 3.0970230102539062 s
Brain elapsed: 5.338260650634766
Spinalcord
sagittal time: 0.12816548347473145 s
coronal time: 0.14362144470214844 s
Spinalcord elapsed: 0.27375149726867676
total voi contours calc time: 39.57941555976868 s
```

now:
```
voi: External, elapsed 4.589036226272583 s
voi: CTV_6000, elapsed 0.377030611038208 s
voi: BrainStem, elapsed 0.11072230339050293 s
voi: Chiasm, elapsed 0.03294968605041504 s
voi: LacrimalGland_L, elapsed 0.02789473533630371 s
voi: LacrimalGland_R, elapsed 0.02592182159423828 s
voi: Lens_L, elapsed 0.013965129852294922 s
voi: Lens_R, elapsed 0.01599907875061035 s
voi: OpticNerve_L, elapsed 0.07480025291442871 s
voi: OpticNerve_R, elapsed 0.07181715965270996 s
voi: CTV_5000, elapsed 0.5401036739349365 s
voi: Brain, elapsed 0.9055707454681396 s
voi: Spinalcord, elapsed 0.09624505043029785 s
total elapsed 6.892634630203247 s
```

### CODE REVIEW

File: pytrip/lib/core.c [hunk 1]
```diff
@@ -751,6 +751,61 @@ static PyObject * slice_on_plane(PyObject *self, PyObject *args)
         }
     }
 
+    // choose intersection plane type
```
Comment (grzanka) [inline pytrip/lib/core.c:754]: Can't this be compressed into single loop with one more pass and some extra `if` statement, like:

```
if( i < length) {
 first_index = i;
 second_index = i+1;
} else {
 first_index = length;
 second_index = 0;
} 

Also - add a comment mentioning that this is needed to close the loop.

File: pytrip/vdx.py [hunk 2]
```diff
@@ -923,10 +923,7 @@ def get_2d_slice(self, plane, depth):
         for _slice in self.slices:  # TODO: slices must be sorted first, but wouldn't they always be ?
 
             # thanks to previous call to `concat_contour` so there is exactly one contour in each slice
-            contour = np.array(_slice.contours[0].contour)
-            # we need to add first point at the end, without it slice_on_plane won't work properly
-            #   because it operates on closed chains of points
-            contour = np.append(contour, [contour[0]], axis=0)
+            contour = _slice.contours[0].contour
```
Comment (grzanka) [inline pytrip/vdx.py:926]: Was it this place where the time saving occurred ?

File: pytrip/lib/core.c [hunk 3]
```diff
@@ -822,6 +877,129 @@ static PyObject* function_ranges(PyObject *self, PyObject *args){
 
 }
 
+static PyObject* binary_search_intersection(PyObject *self, PyObject *args){
```
Comment (michal367) [inline pytrip/lib/core.c:880]: add comment what this function does

Comment (michal367) [reply pytrip/lib/core.c:880]: and also comment function `function_ranges`

File: pytrip/vdx.py (hunk 2 above)
Comment (rozchlastywacz) [reply pytrip/vdx.py:926]: creating numpy array took almost all the time of calculating intersection

File: pytrip/lib/core.c (hunk 1 above)
Comment (rozchlastywacz) [reply pytrip/lib/core.c:754]: I think it can worsen efficiency, at the end of the for loop we know that the loop has ended, we don't need to check that in every step of loop as in Your suggestion

File: pytrip/lib/core.c (hunk 3 above)
Comment (rozchlastywacz) [reply pytrip/lib/core.c:880]: no.

File: pytrip/vdx.py (hunk 2 above)
Comment (grzanka) [reply pytrip/vdx.py:926]: nice catch !
Was it due to the fact, that array was relatively small ?

File: pytrip/lib/core.c (hunk 1 above)
Comment (grzanka) [reply pytrip/lib/core.c:754]: Yes, but it improved code readability, as now we have huge DYI violation by large block being copy-pasted

Comment (rozchlastywacz) [reply pytrip/lib/core.c:754]: fast code >>> good looking code :)
but i will test that suggestions

Comment (rozchlastywacz) [reply pytrip/lib/core.c:754]: did you mean DRY? Don't repeat yourself?

File: pytrip/vdx.py (hunk 2 above)
Comment (rozchlastywacz) [reply pytrip/vdx.py:926]: creating numpy array was a redundant operation, but using it made C code look cleaner because of numpy C API. 
We changed C code to use operations on classic python arrays, so creating numpy array was not necessary, also instead of adding first point to the end of contour array, we added another for loop body after for loop just for checking last and first point

tl; dr:
creating numpy array wasnt needed and took almost 99 percent of intersecting time

File: pytrip/lib/core.c (hunk 1 above)
Comment (grzanka) [reply pytrip/lib/core.c:754]: yes

you can also calculate index with modulo division:

```
first_index = i % length
second_index = (i+1) % length
```

File: pytrip/lib/core.c [hunk 4]
```diff
@@ -691,26 +692,25 @@ static PyObject * slice_on_plane(PyObject *self, PyObject *args)
 
     // loop over all line segments in the input chain of points
     // if input chain has only one element this loop won't be executed
-    for(i = 0; i < PyArray_DIM(vec_slice, 0)-1; i++)
+    length = PyList_Size(vec_slice)-1;
```
Comment (michal367) [inline pytrip/lib/core.c:695]: it's not length, because you substract 1

File: pytrip/vdx.py [hunk 5]
```diff
@@ -917,25 +917,35 @@ def get_2d_slice(self, plane, depth):
 
         # a list to collect all intersections found in all loop passes
         all_intersections = []
+        magic_length = 50
```
Comment (grzanka) [inline pytrip/vdx.py:920]: what about `switch_to_binary_search_threshold` ?

**Comment by rozchlastywacz** [comment]
```
binary search

voi: External, elapsed 3.2484934329986572 s
voi: Bladder, elapsed 0.34410786628723145 s
voi: PenileBulb, elapsed 0.034906625747680664 s
voi: FemoralHead_L, elapsed 0.14261770248413086 s
voi: FemoralHead_R, elapsed 0.2493736743927002 s
voi: Rectum, elapsed 0.21650290489196777 s
voi: iCTVp_04_6000, elapsed 0.14586639404296875 s
voi: Rectalwall, elapsed 0.30973076820373535 s
voi: iCTVp+sb_4800, elapsed 0.19720792770385742 s
voi: Ref, elapsed 0.0030045509338378906 s
total elapsed 4.9147093296051025 s

voi: duodenum, elapsed 0.6756901741027832 s
voi: kidney_L, elapsed 0.5653185844421387 s
voi: kidney_R, elapsed 0.5904614925384521 s
voi: liver, elapsed 1.673318862915039 s
voi: pancreas, elapsed 0.4058980941772461 s
voi: spleen, elapsed 0.5664994716644287 s
voi: Spinalcord, elapsed 0.29897093772888184 s
voi: stomach, elapsed 0.703667402267456 s
voi: iCTV_5040, elapsed 0.38197851181030273 s
voi: iCTV_5580, elapsed 0.2522592544555664 s
voi: Ref, elapsed 0.0030291080474853516 s
voi: External, elapsed 3.976346969604492 s
total elapsed 10.102450132369995 s
```
```
linear search

voi: External, elapsed 4.004110097885132 s
voi: Bladder, elapsed 0.3600771427154541 s
voi: PenileBulb, elapsed 0.034906864166259766 s
voi: FemoralHead_L, elapsed 0.14262127876281738 s
voi: FemoralHead_R, elapsed 0.14759278297424316 s
voi: Rectum, elapsed 0.20450735092163086 s
voi: iCTVp_04_6000, elapsed 0.13564300537109375 s
voi: Rectalwall, elapsed 0.4699103832244873 s
voi: iCTVp+sb_4800, elapsed 0.1924901008605957 s
voi: Ref, elapsed 0.0019555091857910156 s
total elapsed 5.69935941696167 s

voi: duodenum, elapsed 0.6545774936676025 s
voi: kidney_L, elapsed 0.5595252513885498 s
voi: kidney_R, elapsed 0.5341627597808838 s
voi: liver, elapsed 1.7050607204437256 s
voi: pancreas, elapsed 0.4287686347961426 s
voi: spleen, elapsed 0.40099000930786133 s
voi: Spinalcord, elapsed 0.514641523361206 s
voi: stomach, elapsed 0.56947922706604 s
voi: iCTV_5040, elapsed 0.2593080997467041 s
voi: iCTV_5580, elapsed 0.21346664428710938 s
voi: Ref, elapsed 0.002001523971557617 s
voi: External, elapsed 4.479854345321655 s
total elapsed 10.331799507141113 s
```
```binary if len(contour) > 50 else linear

voi: External, elapsed 3.6747403144836426 s
voi: Bladder, elapsed 0.49168825149536133 s
voi: PenileBulb, elapsed 0.06383132934570312 s
voi: FemoralHead_L, elapsed 0.22048211097717285 s
voi: FemoralHead_R, elapsed 0.15608906745910645 s
voi: Rectum, elapsed 0.23301434516906738 s
voi: iCTVp_04_6000, elapsed 0.1396348476409912 s
voi: Rectalwall, elapsed 0.29243040084838867 s
voi: iCTVp+sb_4800, elapsed 0.18539118766784668 s
voi: Ref, elapsed 0.001993894577026367 s
total elapsed 5.484140157699585 s

voi: duodenum, elapsed 0.5655210018157959 s
voi: kidney_L, elapsed 0.43078017234802246 s
voi: kidney_R, elapsed 0.4647223949432373 s
voi: liver, elapsed 1.3007078170776367 s
voi: pancreas, elapsed 0.24189496040344238 s
voi: spleen, elapsed 0.26429104804992676 s
voi: Spinalcord, elapsed 0.24135303497314453 s
voi: stomach, elapsed 0.5096778869628906 s
voi: iCTV_5040, elapsed 0.2658402919769287 s
voi: iCTV_5580, elapsed 0.23139572143554688 s
voi: Ref, elapsed 0.001997709274291992 s
voi: External, elapsed 3.9494316577911377 s
total elapsed 8.474636316299438 s
```

**Comment by rozchlastywacz** [comment]
image

same test launched twice, on the same machine, same app, nothing changed, except for order of loaded files:

first launch:
```
binary if len(contour) > 50 else linear
if binary:
  binary if binary_search_range > 50 else linear

voi: External, elapsed 3.1879823207855225 s
voi: Bladder, elapsed 0.4182724952697754 s
voi: PenileBulb, elapsed 0.05209183692932129 s
voi: FemoralHead_L, elapsed 0.18416857719421387 s
voi: FemoralHead_R, elapsed 0.13272380828857422 s
voi: Rectum, elapsed 0.18000411987304688 s
voi: iCTVp_04_6000, elapsed 0.10998654365539551 s
voi: Rectalwall, elapsed 0.24477267265319824 s
voi: iCTVp+sb_4800, elapsed 0.15303611755371094 s
voi: Ref, elapsed 0.0 s
total elapsed 4.710136413574219 s

voi: duodenum, elapsed 0.3991119861602783 s
voi: kidney_L, elapsed 0.40128469467163086 s
voi: kidney_R, elapsed 0.40085625648498535 s
voi: liver, elapsed 1.3769521713256836 s
voi: pancreas, elapsed 0.30296897888183594 s
voi: spleen, elapsed 0.31831812858581543 s
voi: Spinalcord, elapsed 0.25022435188293457 s
voi: stomach, elapsed 0.5239012241363525 s
voi: iCTV_5040, elapsed 0.27240800857543945 s
voi: iCTV_5580, elapsed 0.20270037651062012 s
voi: Ref, elapsed 0.0 s
voi: External, elapsed 4.1474363803863525 s
total elapsed 8.611509084701538 s
```
second launch, faster one:
```
binary if len(contour) > 50 else linear
if binary:
  binary if binary_search_range > 50 else linear

voi: External, elapsed 2.5636441707611084 s
voi: Bladder, elapsed 0.2939891815185547 s
voi: PenileBulb, elapsed 0.030646562576293945 s
voi: FemoralHead_L, elapsed 0.12221169471740723 s
voi: FemoralHead_R, elapsed 0.12476181983947754 s
voi: Rectum, elapsed 0.16330504417419434 s
voi: iCTVp_04_6000, elapsed 0.1119847297668457 s
voi: Rectalwall, elapsed 0.25495123863220215 s
voi: iCTVp+sb_4800, elapsed 0.16068315505981445 s
voi: Ref, elapsed 0.0 s
total elapsed 3.8270554542541504 s

voi: duodenum, elapsed 0.3780391216278076 s
voi: kidney_L, elapsed 0.3394308090209961 s
voi: kidney_R, elapsed 0.3353567123413086 s
voi: liver, elapsed 1.3653676509857178 s
voi: pancreas, elapsed 0.2877640724182129 s
voi: spleen, elapsed 0.3064582347869873 s
voi: Spinalcord, elapsed 0.2019953727722168 s
voi: stomach, elapsed 0.41782331466674805 s
voi: iCTV_5040, elapsed 0.22156357765197754 s
voi: iCTV_5580, elapsed 0.17237448692321777 s
voi: Ref, elapsed 0.0 s
voi: External, elapsed 3.6047849655151367 s
total elapsed 7.6392621994018555 s
```

**Comment by rozchlastywacz** [comment]
```
voi: duodenum, elapsed 0.573498010635376 s
voi: kidney_L, elapsed 0.4527902603149414 s
voi: kidney_R, elapsed 0.4228696823120117 s
voi: liver, elapsed 1.4157230854034424 s
voi: pancreas, elapsed 0.2891712188720703 s
voi: spleen, elapsed 0.27219605445861816 s
voi: Spinalcord, elapsed 0.2520003318786621 s
voi: stomach, elapsed 0.5605316162109375 s
voi: iCTV_5040, elapsed 0.24590253829956055 s
voi: iCTV_5580, elapsed 0.20545077323913574 s
voi: Ref, elapsed 0.002992391586303711 s
voi: External, elapsed 3.642102003097534 s
total elapsed 8.358217477798462 s

voi: duodenum, elapsed 0.43488049507141113 s
voi: kidney_L, elapsed 0.3570098876953125 s
voi: kidney_R, elapsed 0.5206093788146973 s
voi: liver, elapsed 1.2199831008911133 s
voi: pancreas, elapsed 0.3181757926940918 s
voi: spleen, elapsed 0.35884904861450195 s
voi: Spinalcord, elapsed 0.2275843620300293 s
voi: stomach, elapsed 0.4732096195220947 s
voi: iCTV_5040, elapsed 0.2903590202331543 s
voi: iCTV_5580, elapsed 0.2444009780883789 s
voi: Ref, elapsed 0.0029914379119873047 s
voi: External, elapsed 4.318772077560425 s
total elapsed 8.77056360244751 s

voi: External, elapsed 3.178631544113159 s
voi: Bladder, elapsed 0.3207540512084961 s
voi: PenileBulb, elapsed 0.030943632125854492 s
voi: FemoralHead_L, elapsed 0.13332271575927734 s
voi: FemoralHead_R, elapsed 0.13264918327331543 s
voi: Rectum, elapsed 0.19029688835144043 s
voi: iCTVp_04_6000, elapsed 0.12277722358703613 s
voi: Rectalwall, elapsed 0.27609896659851074 s
voi: iCTVp+sb_4800, elapsed 0.16927719116210938 s
voi: Ref, elapsed 0.0019948482513427734 s
total elapsed 4.566635370254517 s

voi: External, elapsed 3.2810897827148438 s
voi: Bladder, elapsed 0.3323841094970703 s
voi: PenileBulb, elapsed 0.03091144561767578 s
voi: FemoralHead_L, elapsed 0.13065552711486816 s
voi: FemoralHead_R, elapsed 0.13289642333984375 s
voi: Rectum, elapsed 0.19522714614868164 s
voi: iCTVp_04_6000, elapsed 0.1217200756072998 s
voi: Rectalwall, elapsed 0.2661888599395752 s
voi: iCTVp+sb_4800, elapsed 0.16455984115600586 s
voi: Ref, elapsed 0.0029914379119873047 s
total elapsed 4.663631916046143 s

```
times for current version

File: pytrip/lib/core.c [hunk 6]
```diff
@@ -641,12 +644,62 @@ static PyObject * calculate_lvh_slice(PyObject *self, PyObject *args)
     return PyArray_Return(vec_out);
 }
 
+static void append_intersection(PyObject *list_out, double x, double y, double z){
```
Comment (grzanka) [inline pytrip/lib/core.c:647]: I think this should be rather called `append_point_to_list`
The method is not specific to intersection

File: pytrip/lib/core.c [hunk 7]
```diff
@@ -26,6 +26,9 @@
 #include "numpy/arrayobject.h"
 #include "structmember.h"
 
+// macro to ease python array manipulation
```
Comment (grzanka) [inline pytrip/lib/core.c:29]: Explain in the comment what kind of data this macro expects

File: pytrip/lib/core.c [hunk 8]
```diff
@@ -641,12 +644,62 @@ static PyObject * calculate_lvh_slice(PyObject *self, PyObject *args)
...
+    list_item = PyList_New(3);
+
+    PyList_SetItem(list_item, 0, PyFloat_FromDouble(x));
+    PyList_SetItem(list_item, 1, PyFloat_FromDouble(y));
+    PyList_SetItem(list_item, 2, PyFloat_FromDouble(z));
+
+    PyList_Append(list_out, list_item);
+}
+
+static void calc_intersect_sagittal(PyObject *vec_slice, PyObject *list_out, double depth, int i_0, int i_1){
```
Comment (grzanka) [inline pytrip/lib/core.c:659]: add a comment explaining which plane is sagittal - XY, YZ or XZ ?

File: pytrip/lib/core.c [hunk 9]
```diff
@@ -641,12 +644,62 @@ static PyObject * calculate_lvh_slice(PyObject *self, PyObject *args)
...
+    double x_0, x_1, y_0, y_1, z;
+    double a;
```
Comment (grzanka) [inline pytrip/lib/core.c:661]: rename a for fraction or factor, so it resembles linear interpolation

File: pytrip/lib/core.c [hunk 10]
```diff
@@ -641,12 +644,62 @@ static PyObject * calculate_lvh_slice(PyObject *self, PyObject *args)
...
+
+        append_intersection(list_out, depth, ((depth-x_0)*a+y_0), z);
+    }   
+}
+
+static void calc_intersect_coronal(PyObject *vec_slice, PyObject *list_out, double depth, int i_0, int i_1){
+    double x_0, x_1, y_0, y_1, z;
+    double a;
+    
+    y_0 = GET_COORDINATE(vec_slice, i_0, 1);
```
Comment (grzanka) [inline pytrip/lib/core.c:685]: add similar comments as in calc_intersect_sagittal

File: pytrip/lib/core.c (hunk 8 above)
Comment (grzanka) [reply pytrip/lib/core.c:659]: also explain what this method is doing, explaining what is the meaning of parameters

File: pytrip/lib/core.c (hunk 9 above)
Comment (rozchlastywacz) [reply pytrip/lib/core.c:661]: i think it does:
image

File: pytrip/vdx.py [hunk 11]
```diff
@@ -1427,6 +1449,8 @@ def __init__(self, cube=None, plane=None):
         # added to make this class more generic
         # now it stores slices in sagittal and coronal
         self._plane = plane
+        self.ranges_cor = None
+        self.ranges_sag = None
```
Comment (grzanka) [inline pytrip/vdx.py:1453]: What are these flags ?

**Review by grzanka — APPROVED** [review]
LGTM
